kreait/firebase-php
# Rework credentials discovery

Addresses #399 and #400 (hopefully)

### Changes

- The auto-discovery of credentials to be used for authentication against Google's APIs is now handled by the CredentialsLoader/ApplicationDefaultCredentials component of the `google/auth` library
- Added possibility to specify a project ID to configure the Firebase Factory, this _should_ make it possible to use one credential (usually a service account) to access multiple projects:

```php
$firstProjectFactory = (new Factory)->withProjectId('my-first-project');
$secondProjectFactory = (new Factory)->withProjectId('my-second-project');
```

### How to test

Update the dependency on the SDK in your `composer.json` to something like this:

```
"require": {
	// ...
	"kreait/firebase-php": "dev-credentials-discovery as 4.42.0"
    // ...
}

```

It might be necessary to also add

```
{
    "minimum-stability": "dev",
    "prefer-stable": true
}
```

- Authenticating with a Service Account should work the same as before
- Authenticating with user refresh credentials should be possible, but will result in reduced functionality. If you use user refresh credentials (e.g. after `gcloud auth application-default login`), you need to set the environment variable `SUPPRESS_GCLOUD_CREDS_WARNING` to `true`. You can do this with
  - `export SUPPRESS_GCLOUD_CREDS_WARNING=true` in your terminal
  - Prefixing a script execution: `SUPPRESS_GCLOUD_CREDS_WARNING=true php my-script.php`
  - Adding `SUPPRESS_GCLOUD_CREDS_WARNING=true` to `.env` (or similar) when using a Framework that supports it.
  - Adding `putenv('SUPPRESS_GCLOUD_CREDS_WARNING=true');` during the bootstrap process of you application.

### Todo


:octocat:
